src/events/deposits.upsert.service.ts

```diff
@@ -288,4 +228,190 @@ export class DepositsUpsertService {
+                depositParams.push({
+                  graffiti,
+                  amount,
+                  transaction_hash: standardizeHash(transaction.hash),
+                  block_hash: blockHash,
+                  block_sequence: operation.block.sequence,
+                  main: true,
+                  network_version: networkVersion,
+                });
+              }
+            }
+
+            // Bulk load unique users and map by graffiti
+            users = await this.usersService.findManyAndMapByGraffiti(
+              depositParams.map((d) => d.graffiti),
+            );
+
+            // Filter deposits made by unknown users
+            depositParams = depositParams.filter((d) => users.has(d.graffiti));
+
+            // Deposits are shared between blocks, so we need to reassign all the ones on other blocks
+            await prisma.deposit.updateMany({
+              data: {
+                block_hash: blockHash,
+                main: true,
+              },
+              where: {
+                AND: depositParams.map((deposit) => ({
+                  transaction_hash: deposit.transaction_hash,
+                  graffiti: deposit.graffiti,
+                })),
+                network_version: networkVersion,
+              },
+            });
+
+            // Now create new not existing deposits
+            await prisma.deposit.createMany({
+              data: depositParams,
+              skipDuplicates: true,
+            });
+
+            deposits = await prisma.deposit.findMany({
+              where: {
+                block_hash: blockHash,
+                network_version: networkVersion,
+              },
+            });
+
+            const eventPayloads = [];
+            const usersFiltered = new Map<string, User>();
+
+            // Create the event payloads filtering events and users on points < 0
+            for (const deposit of deposits) {
+              const points = this.eventsService.calculateDepositPoints(deposit);
+
+              if (points <= 0) {
+                continue;
+              }
+
+              const user = users.get(deposit.graffiti);
+
+              // This should NEVER happen but can happen if a user is deleted after they made a deposit
+              if (!user) {
+                continue;
+              }
+
+              usersFiltered.set(deposit.graffiti, user);
+
+              eventPayloads.push({
+                occurred_at: blockTimestamp.toISOString(),
+                type: EventType.SEND_TRANSACTION,
+                user_id: user.id,
+                points: points,
+                deposit_id: deposit.id,
+              });
+            }
+
+            users = usersFiltered;
+
+            // This SHOULD NOT be needed but this cleans up bad data in the server and crashes without this
+            await prisma.event.deleteMany({
+              where: {
+                deposit_id: {
+                  in: deposits.map((deposit) => deposit.id),
```

Comment:
nit - we can gather deposit IDs in the loop above instead of iterating again

Reply:
changed